server.rb

```diff
@@ -1,163 +1,185 @@
+    @data[params[:channel].to_sym] || halt(404)
+  end
+end
+
+def serve_file(path)
+  send_file File.join settings.public_folder, path
+end
+
+class Hash
+  def add_age(time)
+    merge age: (time - self[:serverseconds])
+  end
+end
+
+DATA = Channels.new(INIT_STATE)
+
+
+get '/channels' do
+  DATA.channels
+end
 
 get '/getdata/:channel' do
-	halt(404) if @@data[params[:channel].to_sym].nil?
-	data = @@data[params[:channel].to_sym][:routes].values.map {|x| x.map {|n| n.merge({:age => (thetime-n[:serverseconds])})} }.to_json
-	if request["callback"]
-		content_type 'text/plain'
-		"#{request["callback"]}(#{data})"
-	else
-		content_type :json
-		data
-	end
+  DATA.respond params do |time, channel_data|
+    channel_data[:routes].map do |k, v|
+      [k, v.add_age(time)]
+    end
+  end
 end
 
 get '/gethistory/:channel' do
-	halt(404) if @@data[params[:channel].to_sym].nil?
-
-	thetime = Time.now.utc.to_i # time in SECONDS
-	#ugly insertion of age
-	data = @@data[params[:channel].to_sym][:history].values.map {|x| x.map {|n| n.merge({:age => (thetime-n[:serverseconds])})} }.to_json
-	if request["callback"]
-		content_type 'text/plain'
-		"#{request["callback"]}(#{data})"
-	else
-		content_type :json
-		data
-	end
+  # ugly insertion of age
+  DATA.respond params do |time, channel_data|
+    channel_data[:history].map do |k, v|
+      newv = v.map do |n|
+        n.add_age(time)
+      end
+      [k, newv]
+    end
+  end
 end
 
 post '/post/:channel' do
-	halt(404) if @@data[params[:channel].to_sym].nil?
-	return if params[:accuracy].to_f > 100
-
-	@@data_lock.synchronize{
-		thetime = Time.now.utc.to_i # time in SECONDS
-		#@@data[params[:channel].to_sym] ||= {routes:{}, info:{}} # we'll deal with initialisation elsewhere
-		reading = {
-			route: params[:route],
-			longitude: params[:longitude],
-			latitude: params[:latitude],
-			timestamp: params[:timestamp],
-			last_restarted: params[:last_restarted],
-			accuracy: params[:accuracy],
-			serverseconds: thetime
-		}
-
-		#DATA
-		@@data[params[:channel].to_sym][:routes][params[:route]] = reading
-
-		#HISTORY
-		#set to empty array if not exist...
-		@@data[params[:channel].to_sym][:history][params[:route]] ||= []
-		@@data[params[:channel].to_sym][:history][params[:route]].unshift(reading)
-
-		#TODO: limit the size of the history array. the stuff below crashes the server...
-		l = @@data[params[:channel].to_sym][:history][params[:route]].length
-		if (l > @@max_history) 
-			@@data[params[:channel].to_sym][:history][params[:route]] = (@@data[params[:channel].to_sym][:history][params[:route]]).take(@@max_history)
-		end 
-	}
-end
-
-get '/displayJS' do
-  send_file File.join(settings.public_folder, 'bustracker.html')
+  DATA.respond params do |time, channel_data|
+    unless params[:accuracy].to_f > 100
+
+      route_id = params[:route]
+
+      reading = {
+        longitude: params[:longitude],
+        latitude: params[:latitude],
+        timestamp: params[:timestamp],
```

Comment:
Not sure whether to get rid of timestamp field. It is potentially useful for dealing with out of order messages, but as it stands we are not parsing it as a date.

It would be more useful if the date were in ISO-8601 format, but that would require changing the sending application.